engine/src/main/java/org/pentaho/di/trans/steps/script/ScriptAddedFunctions.java

```diff
@@ -0,0 +1,2532 @@
+    Object FunctionContext ) {
+    if ( ArgList.length == 1 ) {
+      return Const.getDigitsOnly( (String) ArgList[0] ); // TODO AKRETION ensure
+    } else {
+      throw new RuntimeException( "The function call getDigitsOnly requires 1 argument." );
+
+    }
+  }
+
+  public static boolean LuhnCheck( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+
+    boolean returnCode = false;
+
+    if ( ArgList.length == 1 ) {
+      if ( !isNull( ArgList ) && !isUndefined( ArgList ) ) {
+        try {
+          int sum = 0;
+          int digit = 0;
+          int addend = 0;
+          boolean timesTwo = false;
+          String argstring = (String) ArgList[0];
+
+          for ( int i = argstring.length() - 1; i >= 0; i-- ) {
+            digit = Integer.parseInt( argstring.substring( i, i + 1 ) );
+            if ( timesTwo ) {
+              addend = digit * 2;
+              if ( addend > 9 ) {
+                addend -= 9;
+              }
+            } else {
+              addend = digit;
+            }
+            sum += addend;
+            timesTwo = !timesTwo;
+          }
+
+          int modulus = sum % 10;
+          if ( modulus == 0 ) {
+            returnCode = true;
+          }
+        } catch ( Exception e ) {
+          // No Need to throw exception
+          // This means that input can not be parsed to Integer
+        }
+
+      }
+    } else {
+      throw new RuntimeException( "The function call LuhnCheck requires 1 argument." );
+
+    }
+    return returnCode;
+  }
+
+  public static int indexOf( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+
+    int returnIndex = -1;
+
+    if ( ArgList.length == 2 || ArgList.length == 3 ) {
+      if ( !isNull( ArgList ) && !isUndefined( ArgList ) ) {
+        String string = (String) ArgList[0];
+        String subString = (String) ArgList[1];
+
+        int fromIndex = 0;
+        if ( ArgList.length == 3 ) {
+          fromIndex = (int) Math.round( (Double) ArgList[2] );
+        }
+        returnIndex = string.indexOf( subString, fromIndex );
+      }
+    } else {
+      throw new RuntimeException( "The function call indexOf requires 2 or 3 arguments" );
+    }
+    return returnIndex;
+  }
+
+  public static Object getTransformationName( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+    try {
+      Object objTranName = actualObject.get( "_TransformationName_" );
+      return objTranName;
+    } catch ( Exception e ) {
+      throw new RuntimeException( e.toString() );
+    }
+  }
+
+  public static void appendToFile( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+
+    if ( !isNull( ArgList ) && !isUndefined( ArgList ) ) {
+      try {
+        FileOutputStream file = new FileOutputStream( (String) ArgList[0], true );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *file*. The resource is referenced at lines 205, 206, 207. The resource is closed at line 208. There are other execution paths that do not contain closure statements, for example, when DataOutputStream.flush throws an exception. Ensure *file* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).